### PR TITLE
[Common] Added JIT-compiled fused cast transpose kernels

### PR DESCRIPTION
**Pull request by Oleg-Goncharov**
# Description

This PR extends the JIT-compilation on the fused cast transpose kernels using CUDA Driver API (nvrtc), which enables performance optimizations otherwise not available during runtime, and reduces the size of the TE binary `libtransformer_engine.so` by 4.5% (from 292M to 279M).
The performance was benchmarked for the `cast_transpose_dbias` and `cast_transpose_dbias_dgelu` kernels for the matrix size `2048x12288` and different combinations of input types. The results are provided in the table below:
JIT-kernels

NOTES:
- Original - fused cast transpose kernels before the commit #884 
- Template - current state
- JIT - proposed by this PR
- Benchmarked on H100 HBM3
- Time measured in microseconds

Performance of the kernels remains the same, with some marginal changes (~2-6%) of nvte functions runtime.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Changes

- Modified the logic in the `cast_transpose_fusion.cu`
- Added the corresponding source file into the `transformer_engine/common/transpose/rtc` folder
- Adjusted the CMake for generating a string header from that file

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/common/transpose/cast_transpose_fusion.cu [hunk 1]
```diff
@@ -635,71 +755,121 @@ void cast_transpose_fused(const Tensor &input,
...
-                cudaFuncSetAttribute(
+            // Runtime-compiled tuned kernel
+            if (jit_compiled) {
+                constexpr const char *itype_name  = TypeInfo<InputType>::name;
+                constexpr const char *itype2_name = TypeInfo<InputType2>::name;
+                constexpr const char *otype_name  = TypeInfo<OutputType>::name;
+
+                // Compile NVRTC kernel if needed and launch
+                auto& rtc_manager = rtc::KernelManager::instance();
+                const std::string kernel_label = concat_strings("cast_transpose"
```
Comment (timmoon10) [inline transformer_engine/common/transpose/cast_transpose_fusion.cu:806]: To make sure this key is distinct from https://github.com/NVIDIA/TransformerEngine/blob/89d44e2759f27c690f819e6d48a695081b900b62/transformer_engine/common/transpose/cast_transpose.cu#L317
```suggestion
                const std::string kernel_label = concat_strings("cast_transpose_fused"
```

**Review by timmoon10 — APPROVED** [review]
LGTM

File: transformer_engine/common/transpose/cast_transpose_fusion.cu [hunk 2]
```diff
@@ -568,16 +464,46 @@ cast_transpose_fused_kernel_notaligned(const Param param,
...
+    } else if (OP == &dsilu<ComputeType, ComputeType>) {
+        return 4;
+    } else if (OP == &drelu<ComputeType, ComputeType>) {
+        return 5;
+    } else if (OP == &dsrelu<ComputeType, ComputeType>) {
+        return 6;
+    } else {
+        return 0;
+    }
+}
```
Comment (phu0ngng) [inline transformer_engine/common/transpose/cast_transpose_fusion.cu:505]: Hi. You may want to have a look at the `NVTE_Activation_Type` in `common/include/transformer_engine/activation.h` and see if you can use that enum here.

Comment (Oleg-Goncharov) [reply transformer_engine/common/transpose/cast_transpose_fusion.cu:505]: Hi Phuong, thanks a lot for the review and the comment. I completely agree that enum would be the way to go, but its usage becomes a problem during JIT compilation. The RTC manager converts kernel input parameters into strings and then concatenates them into a single kernel label. This process is more complicated and convoluted on the both sides, i.e., from the calling routine and from the to-be-compiled-code represented as a string. Passing such a enum type became problematic, so I decided to follow the most convenient way with using the mapping from an activation function pointer to an integer number

**Pull request merged**